File: frontend/src/features/analytics/types/analytics.ts
export type AnalyticsPeriod = "7d" | "30d" | "90d" | "12m";

export interface AnalyticsPoint {
  date: string;
  scans: number;
  unique_visitors: number;
}

export interface AnalyticsBreakdown {
  label: string;
  value: number;
  percentage: number;
}

export interface TopQrCode {
  id: string;
  label: string;
  scans: number;
  unique_visitors: number;
}

export interface RecentScanEvent {
  id: string;
  generation_id: string;
  qr_label: string;
  device_type: string;
  browser: string;
  operating_system: string;
  country: string;
  city: string;
  scanned_at: string;
}

export interface AnalyticsOverview {
  period: AnalyticsPeriod;
  starts_at: string;
  ends_at: string;
  total_scans: number;
  unique_visitors: number;
  previous_total_scans: number;
  scan_change_percentage: number | null;
  series: AnalyticsPoint[];
  devices: AnalyticsBreakdown[];
  browsers: AnalyticsBreakdown[];
  operating_systems: AnalyticsBreakdown[];
  countries: AnalyticsBreakdown[];
  cities: AnalyticsBreakdown[];
  top_qr_codes: TopQrCode[];
  recent_scans: RecentScanEvent[];
}

File: backend/app/schemas/analytics.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field


class AnalyticsPoint(BaseModel):
    date: str
    scans: int
    unique_visitors: int


class AnalyticsBreakdown(BaseModel):
    label: str
    value: int
    percentage: float = Field(ge=0, le=100)


class TopQrCode(BaseModel):
    id: str
    label: str
    scans: int
    unique_visitors: int


class RecentScanEvent(BaseModel):
    id: str
    generation_id: str
    qr_label: str
    device_type: str
    browser: str
    operating_system: str
    country: str
    city: str
    scanned_at: datetime


class AnalyticsOverviewResponse(BaseModel):
    period: Literal["7d", "30d", "90d", "12m"]
    starts_at: datetime
    ends_at: datetime
    total_scans: int
    unique_visitors: int
    previous_total_scans: int
    scan_change_percentage: float | None
    series: list[AnalyticsPoint]
    devices: list[AnalyticsBreakdown]
    browsers: list[AnalyticsBreakdown]
    operating_systems: list[AnalyticsBreakdown]
    countries: list[AnalyticsBreakdown]
    cities: list[AnalyticsBreakdown]
    top_qr_codes: list[TopQrCode]
    recent_scans: list[RecentScanEvent]
